pkg/daemon/daemon.go

```diff
@@ -129,6 +142,41 @@ func (dn *Daemon) Run(stop <-chan struct{}) error {
 	return nil
 }
 
+// runOnce pulls the MachineConfig from either a file or remote URL
+// executes one run, and exits.
+// TODO: Revisit the Run/process methods and refactor and unify them with runOnce
+func (dn *Daemon) runOnce() error {
+	var machineConfig *mcfgv1.MachineConfig
+	var err error
+	if strings.HasPrefix("http://", dn.onceFrom) || strings.HasPrefix("https://", dn.onceFrom) {
+		// If we sense a remote URL has been provided then request MC content
+		// from a remote URL and parse it
+		glog.V(2).Infof("Getting machine config content from %s", dn.onceFrom)
+		machineConfig, err = dn.getMachineConfigFromURL(dn.onceFrom)
+		if err != nil {
+			return err
+		}
+	} else if validPath(dn.onceFrom) {
+		// If we sense a local file has been provided parse it
+		absoluteOnceFrom, err := filepath.Abs(filepath.Clean(dn.onceFrom))
+		if err != nil {
+			return err
+		}
+		machineConfig, err = dn.getMachineConfigFromFile(absoluteOnceFrom)
+		if err != nil {
+			return err
+		}
+	} else {
+		// Otherwise return an error as the input format is unsupported
+		return fmt.Errorf("%s is not a path nor url; can not run once", dn.onceFrom)
+	}
+
+	// At this point we have a populated MachineConfig struct for our desired config
+	// and we run update using an empty machineConfig as there is no provided current state.
+	oldConfig := mcfgv1.MachineConfig{}
```

Comment:
@sdemos thoughts on this. Should we try to create an initial MC based on something, or does an empty one work?

Reply:
an empty one will get rejected by `reconcilable` since there are "changes" to the config in places we can't fix.

it would be much better to use something that represents the actual system state. I'm not sure when or where this will be running exactly, will you be able to get it from kube? is it reasonable to expect the user to provide the current config? maybe we can have the current config always available somewhere on disk?

there are a few other possibilities if getting the current config isn't feasible. we could construct the old config from the irreconcilable parts of the new config. we could rip apart the update function so that the part we call always applies configs and we just say "trust us this one works". or we might be able to just pass it the new config for both the old and the new, since we already determined that we need to update, and most of the functionality in update just applies the config directly and doesn't diff. all of those approaches would break anything that does rely on the current configuration, like file deletion (I think that's the only thing right now, but conceivably there could be more in the future).

Reply:
tl;dr is having an accurate current configuration seems like the vastly preferable option if that's at all possible, but you might be able to hack around it.

Reply:
I don't think we will be able to get the config from kubernetes but let me defer to folks who brought up and will be utilizing this use case. I tend to think we can't being that we can't pull the MC from the API itself, but it doesn't hurt to ask :smile: 

@sdodson / @aaronlevy at the point that MCD runs in the `onceFrom` path is the cluster up and serving machine configs?

Reply:
In the case of the bootstrap ignition config, there would be no API/cluster yet (the output of this step would create one). And for bootstrap it's also not a state that we want to "reconcile". The bootstrap assets normally are on a separate throw-away node - but in the BYO case, they are run on one of the control-nodes - but then cleaned up after.

For all other cases of `onceFrom`, it would be retrieving the state from the API/cluster (either bootstrap API or real API).

Reply:
so in the case of bootstrap we just want to blindly apply it, it sounds like. will we be required to delete files for the bootstrap case? at what point are we running on the bootstrap node? why can't we rely on ignition at that point, since if there is no bootstrap api then isn't this just the initial provisioning step that we are intervening in?

sorry if any of these questions are already answered somewhere else, I guess I haven't been keeping up with this feature as well as I thought I was.

Reply:
> will we be required to delete files for the bootstrap case?

We do want to clean up these files after bootstrap completes, but I don't initially think it should be the responsibility of MCD (worry that there's fair amount of complexity here we could punt on a simple post-cleanup step).

> at what point are we running on the bootstrap node? 

In the BYO-RHEL host case, you've booted 3 masters and then we execute MCD with `--once-from=/on/disk/boostrap.ign` on one of them (no api is available yet). The bootstrap process then brings up a bootstrap MCO that will serve the master.ign. On all the master nodes then we would need to execute MCD `--once-from=https://control/plane` which would retrieve those roles. This overlap of MCD runs is where the deletion of bootstrap would get a little wonky (we shouldn't actually drop bootstrap assets until the control-plane is fully functional).

> why can't we rely on ignition at that point, since if there is no bootstrap api then isn't this just the initial provisioning step that we are intervening in?

This is to support RHEL nodes so they're already booted / no ignition involved. We're 'using' the ignition definitions to keep the source of truth for node configuration the same -- even if it's a bit of overloaded behavior.

Reply:
Thanks for the clarification @aaronlevy. We really appreciate it!